e2e/tests/upgrade-test.rs

```diff
@@ -1,36 +1,438 @@
+    match flavor {
+        Flavor::Alpine => match from_version {
+            // Our timescaledev images are deprecated, but they are the best option for an alpine
+            // image at the moment. They also are nicely tagged with the exact extension version.
+            FromVersion::First => format!(
+                "timescaledev/promscale-extension:0.5.0-ts2-pg{}",
+                pg_version
+            ),
+            FromVersion::Prior => format!(
+                "timescaledev/promscale-extension:0.5.4-ts2.7.0-pg{}",
+                pg_version
+            ),
+        },
+        Flavor::HA => match from_version {
+            // The timescaledb-ha docker images aren't tagged with the version of the promscale
+            // extension that they contain, so we need to keep a map of exact image tags (with
+            // patch version) for the first release (with extension 0.5.0) and the prior release.
+            FromVersion::First => match pg_version {
+                PgVersion::V14 => String::from("timescale/timescaledb-ha:pg14.2-ts2.6.1-p5"),
+
+                PgVersion::V13 => String::from("timescale/timescaledb-ha:pg13.6-ts2.6.1-p5"),
+
+                PgVersion::V12 => String::from("timescale/timescaledb-ha:pg12.10-ts2.6.1-p2"),
+            },
+            FromVersion::Prior => match pg_version {
+                PgVersion::V14 => String::from("timescale/timescaledb-ha:pg14.4-ts2.7.1-p2"),
+
+                PgVersion::V13 => String::from("timescale/timescaledb-ha:pg13.7-ts2.7.1-p2"),
+
+                PgVersion::V12 => String::from("timescale/timescaledb-ha:pg12.11-ts2.7.1-p2"),
+            },
+        },
+    }
+}
+
+fn determine_image_flavor(docker_image: &str) -> Flavor {
+    // Note: here we just look at the image name. It would be more accurate to
+    // introspect the actual image, but for now this works.
+    match docker_image.starts_with("local/dev_promscale_extension") {
+        true => Flavor::Alpine,
+        false => match docker_image.ends_with("-alpine") {
+            true => Flavor::Alpine,
+            false => Flavor::HA,
+        },
+    }
+}
+
+fn determine_postgres_version(docker_image: &str) -> Result<PgVersion, String> {
+    // Note: here we just look at the image name. It would be more accurate to
+    // introspect the actual image, but for now this works.
+    let pg_version = Regex::new("pg[0-9]{2}")
+        .unwrap()
+        .find(docker_image)
+        .unwrap()
+        .as_str()
+        .strip_prefix("pg")
+        .unwrap();
+    PgVersion::try_from(pg_version)
+}
+
+/// determines the first, last, and prior versions of the promscale extension available
+fn available_extension_versions(client: &mut Client) -> (Version, Version, Version) {
+    // version 0.1 won't parse correctly
+    // version 0.5.3 and pre-releases of 0.5.3 were pulled due to a bug the broke upgrades
+    let qry = r"select version 
+    from pg_available_extension_versions 
```

Comment:
Shouldn't this be logged?

Reply:
I'm not sure what you mean by "logged" here?

Reply:
No idea why did the comment move up a few lines. It was meant for Err => continue down below at 417

Reply:
Thanks. Done.